File: supler/src/main/scala/org/supler/errors/ValidationScope.scala
package org.supler.errors

import org.json4s.JField
import org.json4s.JsonAST.JString
import org.supler.FieldPath

trait ValidationScope {
  /**
   * @param parentPath Path of the parent of the field for which validation is to be checked.
   */
  def shouldValidate(parentPath: FieldPath, valueMissing: Boolean): Boolean
  def generateJSONData: ValidationScopeJSONData
}

case class ValidationScopeJSONData(name: String, extra: List[JField] = Nil)

object ValidateFilled extends ValidationScope {
  override def shouldValidate(parentPath: FieldPath, valueMissing: Boolean) = !valueMissing
  override def generateJSONData = ValidationScopeJSONData("filled")
}
object ValidateAll extends ValidationScope {
  override def shouldValidate(parentPath: FieldPath, valueMissing: Boolean) = true
  override def generateJSONData = ValidationScopeJSONData("all")
}
object ValidateNone extends ValidationScope {
  override def shouldValidate(parentPath: FieldPath, valueMissing: Boolean) = false
  override def generateJSONData = ValidationScopeJSONData("none")
}
case class ValidateInPath(rootPath: FieldPath) extends ValidationScope {
  override def shouldValidate(parentPath: FieldPath, valueMissing: Boolean) = parentPath.childOf(rootPath)
  override def generateJSONData = ValidationScopeJSONData("path", List(JField("path", JString(rootPath.toString))))
}

File: supler-js/src/validation/ValidationScope.ts
interface ValidationScope {
  shouldValidate(path: string): boolean
}

class _ValidateAll implements ValidationScope {
  shouldValidate(path: string): boolean { return true; }
}
var ValidateAll = new _ValidateAll();

class _ValidateNone implements ValidationScope {
  shouldValidate(path: string): boolean { return false; }
}
var ValidateNone = new _ValidateNone();

class ValidateInPath implements ValidationScope {
  constructor(private rootPath: string) {}

  shouldValidate(path: string): boolean {
    return path && ((path === this.rootPath) ||
      (path.indexOf(this.rootPath + '.') === 0) ||
      (path.indexOf(this.rootPath + '[') === 0));
  }
}

class ValidationScopeParser {
  static fromJson(json: any): ValidationScope {
    switch (json.name) {
      case 'none': return ValidateNone;
      case 'all': return ValidateAll;
      case 'path': return new ValidateInPath(json.path);
      // other validation scopes are unsupported and we fall back to the server validation
      default: return ValidateNone;
    }
  }
}
